packages/simulation/src/convert.ts

```diff
@@ -0,0 +1,571 @@
+  for (let curElem = 0; curElem < numberOfNFAStates; curElem++) {
+    mergedStates[curElem] = [curElem]
+  }
+  // Merge states until no more merges can be made
+  let isMerged = true
+  while (isMerged) {
+    isMerged = false
+    for (let i = 0; i < numberOfNFAStates; i++) {
+      const visitedStates = new Set(mergedStates[i])
+      // Traverse lambda transitions until no more states can be reached
+      for (let j = 0; j < visitedStates.size; j++) {
+        const curState = Array.from(visitedStates)[j]
+        const curTransitions = initialTransitionTable[curState] || []
+        for (const [toState, symbol] of curTransitions) {
+          if (symbol === undefined && !visitedStates.has(toState)) {
+            visitedStates.add(toState)
+          }
+        }
+      }
+      // Merge the visited states
+      const mergedState = Array.from(visitedStates).sort((a, b) => a - b)
+      // If the arrays are not equal then merge
+      if (!(mergedStates[i].length === mergedState.length && mergedStates[i].every((element, index) => element === mergedState[index]))) {
+        mergedStates[i] = mergedState
+        isMerged = true
+      }
+    }
+  }
+
+  // Find the new initial state and final state based on the merged states
+  let foundInitial = false
+  let foundFinals = false
+  for (const key in mergedStates) {
+    if (!foundInitial && mergedStates[key].includes(curInitialState)) {
+      curInitialState = Math.min(...mergedStates[key])
+      foundInitial = true
+    }
+    if (!foundFinals) {
+      for (const curKey in mergedStates) {
+        // Check if any final states are present in mergedStates
+        const mergedFinalStates = curFinalStates.filter(state => mergedStates[curKey].includes(state))
+        if (mergedFinalStates.length > 0) {
+          // Remove merged final states from curFinalStates
+          curFinalStates = curFinalStates.filter(state => !mergedFinalStates.includes(state))
+          // Add lowest state ID of mergedStates with the current key to curFinalStates
+          curFinalStates.push(Math.min(...mergedStates[curKey]))
+        }
+      }
+      foundFinals = true
+    }
+    // Exit loop if both new initial and final states have been found
+    if (foundInitial && foundFinals) {
+      break
+    }
+  }
+
+  // Merge the states in tempTransitionTable and initialTransitionTable based on the mergedStates array. Go top down so that states that are valid states are also
+  // able to transition to the last state in the chain
+  for (let curElem = numberOfNFAStates - 1; curElem >= 0; curElem--) {
+    const mergedWith = mergedStates[curElem][0]
+    for (let i = 1; i < mergedStates[curElem].length; i++) {
```

Comment:
Please look through your for loops again

`i` is only used on the next line to index `mergedStates[curElem]` so this could just be a `for of` loop